File: app/src/main/java/com/abot/project0z/server/game/packets/BasePacket.kt
package com.abot.project0z.server.game.packets

import android.annotation.SuppressLint
import java.io.ByteArrayOutputStream
import java.io.IOException


public class BasePacket {
    var shouldEncrypt: Boolean = true
    private var opcode = 0
    private var shouldBuildHeader = false
    private lateinit var header: ByteArray
    private lateinit var data: ByteArray

    // Encryption
    private var useDispatchKey = false

    constructor(opcode: Int) {
        this.opcode = opcode
    }

    constructor(opcode: Int, buildHeader: Boolean) {
        this.opcode = opcode
        this.shouldBuildHeader = buildHeader
    }

    fun getOpcode(): Int {
        return opcode
    }

    fun setOpcode(opcode: Int) {
        this.opcode = opcode
    }

    fun useDispatchKey(): Boolean {
        return useDispatchKey
    }

    fun setUseDispatchKey(useDispatchKey: Boolean) {
        this.useDispatchKey = useDispatchKey
    }

    fun getHeader(): ByteArray? {
        return header
    }

    fun setHeader(header: ByteArray) {
        this.header = header
    }

    fun shouldBuildHeader(): Boolean {
        return shouldBuildHeader
    }

    fun getData(): ByteArray? {
        return data
    }

    fun setData(data: ByteArray) {
        this.data = data
    }

    fun build(): ByteArray {
        if (getHeader() == null) {
            this.header = ByteArray(0)
        }

        if (getData() == null) {
            this.data = ByteArray(0)
        }

        val baos =
            ByteArrayOutputStream(2 + 2 + 2 + 4 + getHeader()!!.size + getData()!!.size + 2)

        this.writeUint16(baos, opcode)
        this.writeUint16(baos, header.size)
        this.writeUint32(baos, data.size)
        this.writeBytes(baos, header)
        this.writeBytes(baos, data)

        return baos.toByteArray()
    }

    fun writeUint16(baos: ByteArrayOutputStream, i: Int) {
        // Unsigned short
        baos.write(((i ushr 8) and 0xFF).toByte().toInt())
        baos.write((i and 0xFF).toByte().toInt())
    }

    fun writeUint32(baos: ByteArrayOutputStream, i: Int) {
        // Unsigned int (long)
        baos.write(((i ushr 24) and 0xFF).toByte().toInt())
        baos.write(((i ushr 16) and 0xFF).toByte().toInt())
        baos.write(((i ushr 8) and 0xFF).toByte().toInt())
        baos.write((i and 0xFF).toByte().toInt())
    }

    fun writeBytes(baos: ByteArrayOutputStream, bytes: ByteArray?) {
        try {
            baos.write(bytes)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}

File: server/server/packets/BasePacket.py

from enum import Enum
import struct

from server.packets.Opcodes import OpcodeMeta



class PacketType(Enum):
    BASE_PACKET: bytes = b"\x00"
    PROTOCOL_PACKET: bytes = b"\x01"


class BasePacket(metaclass=OpcodeMeta):
    opcode = None
    disabled = False
    
    def __init__(self, opcode: int, build_header: bool = False):
        self.packet_type = PacketType.BASE_PACKET
        self.should_encrypt: bool = True
        self.opcode: int = opcode
        self.should_build_header: bool = build_header
        self.header: bytes = b''
        self.data: bytes = b''
        self.use_dispatch_key: bool = False


    def get_opcode(self): return self.opcode
    def set_opcode(self, opcode: int): self.opcode = opcode

    def get_use_dispatch_key(self): return self.use_dispatch_key
    def set_use_dispatch_key(self, use_dispatch_key: bool): self.use_dispatch_key = use_dispatch_key

    def get_header(self): return self.header
    def set_header(self, header: bytes): self.header = header

    def get_data(self): return self.data
    def set_data(self, data: bytes): self.data = data

    def build(self) -> bytes:
        if self.header is None:
            self.header = b''
        if self.data is None:
            self.data = b''

        buffer = bytearray()

        # Write Packet type (unsigned short, 2 bytes)
        buffer.extend(struct.pack('>H', self.packet_type))

        # Write opcode (unsigned short, 2 bytes)
        buffer.extend(struct.pack('>H', self.opcode))

        # Write header length (unsigned short, 2 bytes)
        buffer.extend(struct.pack('>H', len(self.header)))

        # Write data length (unsigned int, 4 bytes)
        buffer.extend(struct.pack('>I', len(self.data)))

        # Write header bytes
        buffer.extend(self.header)

        # Write data bytes
        buffer.extend(self.data)

        return bytes(buffer)
    


class ProtocolPacket(BasePacket):

    def __init__(self, opcode: int, stage_id: int = 0, build_header: bool = False):
        self.stage_id: int = stage_id
        super().__init__(opcode, build_header)

        self.packet_type = PacketType.PROTOCOL_PACKET

    def build(self):
        if self.header is None:
            self.header = b''
        if self.data is None:
            self.data = b''
        
        header_buffer = bytearray()

        header_buffer.extend(struct.pack('>H', self.stage_id))
        header_buffer.extend(self.header)

        self.set_header(bytes(header_buffer))

        return super().build()